File: alexp/src/main/java/alexp/macrobase/explanation/lofTest/LOF_subspace.py
import numpy as np
from sklearn.neighbors import LocalOutlierFactor
import pandas as pd
import itertools


def run_lof(dataframe, subspace):
    clf = LocalOutlierFactor(n_neighbors=15, algorithm='brute', metric='euclidean', contamination=0.015)
    clf.fit_predict(dataframe)
    x_scores = clf.negative_outlier_factor_
    points_rel_subspaces = {}
    sorted_ind = np.argsort(x_scores)
    for ind in sorted_ind:
        points_rel_subspaces[ind] = '[' + "".join(subspace) + '] ' + str(x_scores[ind]) + '; '
    return points_rel_subspaces


def test_subspace_2_3_4_5(dataframe, elementsToKeep):
    points_of_interest = [172, 183, 184, 207, 220, 245, 315, 323, 477, 510, 577, 654, 704, 723, 754, 765, 781, 824, 975]
    sub_dataframe = dataframe.iloc[:, [2, 3, 4, 5]]
    points_rel_subspaces = run_lof(sub_dataframe, sub_dataframe.columns)

    points_rel_subspaces = dict(itertools.islice(points_rel_subspaces.items(), elementsToKeep))
    print('\nPoints of interest\n')
    print(points_of_interest)

    print('\nPoints with top-' + str(elementsToKeep) + 'scores accompanied by their subspace and score in this '
                                                       'subspace\n')
    for pointId in points_rel_subspaces:
        print(str(pointId) + '=' + points_rel_subspaces[pointId])

    print('\nInliers scored higher than outliers\n')
    print(set(points_rel_subspaces.keys()) - set(points_of_interest))


if __name__ == '__main__':
    df = pd.read_csv('synth_multidim_010_000.csv')
    test_subspace_2_3_4_5(df, 5)

File: alexp/src/main/java/alexp/macrobase/explanation/lofTest/LOF_subspace.java
package alexp.macrobase.explanation.lofTest;

import alexp.macrobase.ingest.Uri;
import alexp.macrobase.outlier.lof.bkaluza.LOF;
import alexp.macrobase.pipeline.Pipelines;
import alexp.macrobase.pipeline.benchmark.config.AlgorithmConfig;
import alexp.macrobase.pipeline.benchmark.config.BenchmarkConfig;
import alexp.macrobase.pipeline.benchmark.config.DatasetConfig;
import alexp.macrobase.pipeline.config.StringObjectMap;
import com.google.common.base.Joiner;
import com.google.common.collect.Sets;
import edu.stanford.futuredata.macrobase.analysis.classify.Classifier;
import edu.stanford.futuredata.macrobase.datamodel.DataFrame;
import edu.stanford.futuredata.macrobase.datamodel.Schema;
import javafx.util.Pair;
import org.omg.CORBA.INTERNAL;

import java.io.IOException;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

public class LOF_subspace {

    private static String scoreColumn = "_OUTLIER";

    private static String datasetPath = "alexp/data/explanation/TestLOF/synth_010_000_2_3_4_5.yaml";

    private static List<Pair<Integer, String>> runClassifier(DataFrame df, String[] columns, AlgorithmConfig conf) throws Exception {
        Classifier classifier = Pipelines.getClassifier(conf.getAlgorithmId(), conf.getParameters(), columns, null);
        return getPointsScoresInRelSubspaces(df, Arrays.asList(columns), classifier);
    }

    private static List<Pair<Integer, String>> getPointsScoresInRelSubspaces(DataFrame df, List<String> relSubspace, Classifier classifier) throws Exception {
        List<Pair<Integer, String>> pointsRelSubspaces = new ArrayList<>();
        classifier.process(df);
        DataFrame results = classifier.getResults();
        List<Pair<Integer, Double>> pointsScores = new ArrayList<>();
        for(int i=0; i < results.getNumRows(); i++) {
            pointsScores.add(new Pair<>(i, results.getRow(i).getAs(scoreColumn)));
        }
        pointsScores.sort(Comparator.comparing(Pair::getValue));
        Collections.reverse(pointsScores);
        for(Pair<Integer, Double> pair : pointsScores){
            String relSubspaceStr = subspaceToString(relSubspace, pair.getValue());
            pointsRelSubspaces.add(new Pair<>(pair.getKey(), relSubspaceStr));
        }
        return pointsRelSubspaces;
    }

    private static String subspaceToString(List<String> subspaceFeatures, double score) {
        return "[" + Joiner.on(" ").join(subspaceFeatures) + "] " + score + ";";
    }

    private static DataFrame loadDataFrame(String datasetFilePath, String[] features) throws Exception {
        Map<String, Schema.ColType> colTypes = Arrays.stream(features)
                .collect(Collectors.toMap(Function.identity(), c -> Schema.ColType.DOUBLE));

        List<String> requiredColumns = new ArrayList<>(colTypes.keySet());

        return Pipelines.loadDataFrame(new Uri("csv://"+datasetFilePath), colTypes, requiredColumns, null);
    }

    private static void validator(HashMap<Integer, String> pointsRelSubspaces) {
        System.out.println("Print points accompanied by their subspace and score in this subspace");
        for(int pointId : pointsRelSubspaces.keySet()) {
            System.out.println(pointId + " -> " + pointsRelSubspaces.get(pointId));
        }

//        System.out.println("\nInliers scored higher than outliers");
//        pointsRelSubspaces.keySet().removeAll(pointsOfInterest);
//
    }

    private static void test_subspace_2_3_4_5(AlgorithmConfig classifierConf, DatasetConfig datasetConfig, int elementsToKeep) throws Exception{

        HashSet<Integer> pointsOfInterest = Sets.newHashSet(172, 183, 184, 207, 220, 245, 315, 323, 477, 510, 577, 654, 704, 723, 754, 765, 781, 824, 975);

        DataFrame df = loadDataFrame(datasetConfig.getDatasetId(), datasetConfig.getMetricColumns());

        List<Pair<Integer, String>> pointsRelSubspaces = runClassifier(df, datasetConfig.getMetricColumns(), classifierConf);

        System.out.println("\nPoints of interest\n");
        System.out.println(pointsOfInterest);

        System.out.println("\nPoints with top-" + elementsToKeep + " scores accompanied by their subspace and score in this subspace\n");
        for(int i =0; i < elementsToKeep; i++) {
            System.out.println(pointsRelSubspaces.get(i));
        }

        List<Integer> topK = pointsRelSubspaces.subList(0, elementsToKeep)
                .stream()
                .map(Pair::getKey)
                .limit(elementsToKeep)
                .collect(Collectors.toList());

        System.out.println("\nInliers scored higher than outliers\n");
        HashSet<Integer> inliersInTopK = new HashSet<>(topK);
        inliersInTopK.removeAll(pointsOfInterest);
        System.out.println(inliersInTopK);
    }

    public static void main(String[] args) throws Exception {
        BenchmarkConfig bc = BenchmarkConfig.load(StringObjectMap.fromYamlFile(datasetPath));
        for(AlgorithmConfig classifierConf : bc.getClassifierConfigs())
            test_subspace_2_3_4_5(classifierConf, bc.getDatasetConfig(), 5);
    }



}
